rtCommon/bidsStructures.py

```diff
@@ -0,0 +1,255 @@
+"""-----------------------------------------------------------------------------
+
+bidsDataset.py (Create: 10/25/20)
+
+This script implements data structures for representing BIDS-formatted
+datasets.
+
+-----------------------------------------------------------------------------"""
+
+from bids import BIDSLayout
+import bids.config as bc
+from functools import lru_cache
+import logging
+import nibabel as nib
+import os
+import re
+
+logger = logging.getLogger(__name__)
+
+# Silence future warning
+bc.set_option('extension_initial_dot', True)
+
+
+class BidsDataset:
+    def __init__(self, path: str):
+        self.name = os.path.basename(os.path.normpath(path))
+        logger.debug("Loading dataset \"%s\" from: %s", self.name, path)
+        self.data = BIDSLayout(path)
+        logger.debug("Dataset info: %s", self.data)
+
+    def insert(self) -> None:
+        # Insert some type of data to the dataset
+        pass
+
+    def erase(self) -> None:
+        # Erase some type of data from the dataset
+        pass
+
+    def exists(self, path: str) -> bool:
+        if not self.data:
+            logger.debug("Ran BidsDataset.find on a non-initialized dataset")
+            return False
+
+        # search for matching BIDS file in the dataset
+        result = self.data.get_file(path)
+        if result is not None:
+            return True
+
+        # search for matching directory in the dataset
+        abs_path = os.path.join(self.data.root, path)
+        return os.path.isdir(abs_path)
+
+
+class BidsArchive:
+    """
+    Represents a BIDS Archive
+    """
+    def __init__(self, rootPath: str, datasetName: str = None):
+        self.dataset = BidsDataset(rootPath)
+
+    def pathExists(self, path: str) -> bool:
+        """
+        Check whether the provided path is valid within the archive.
+        """
+        return self.dataset.exists(path)
+
+
+class BidsIncremental:
+    """
+    BIDS Incremental data format suitable for streaming BIDS Archives
+    """
+    def __init__(self, niftiImg, imgMetadata, datasetMetadata=None):
```

Comment:
Perhaps separate public and private metadata, with public required and private optional? privateMetaData=None. Or both could be required but stored separately to make anonymizing easier if requested.